### ISSUE TITLE
Add JavaScript interop completions

### ISSUE THREAD

**Issue by bpringe** [clojure-lsp/clojure-lsp#327]
This is just something that would be nice to have in the future. Maybe [clj-suitable](https://github.com/rksm/clj-suitable) or something like it could be used.

**Comment by bpringe** [comment]
Actually I think suitable uses a repl for that... Not sure if it could be used that way from clojure-lsp. Maybe there's some other way though.

**Comment by ericdallo** [comment]
Not sure I understand the issue, why this doesn't work with clojure-lsp already? since clojure-lsp check for the required alias to show completion items?

**Comment by bpringe** [comment]
Does it work for JavaScript object completions? It doesn't seem so to me, at least. https://github.com/rksm/clj-suitable#dynamic-completion-demo

The idea: If I type `(.| js/document)` with `|` marking the position of the cursor, methods and properties of `js/document` will appear — including `querySelector` and `querySelectorAll`.

**Comment by ericdallo** [comment]
Got it, it's completion for java classes in the end, right?
Yeah, that'd be really helpful, but we'd depend on clj-kondo analysis for that

**Comment by bpringe** [comment]
Javascript* methods and properties, in this case. So, different approach than for Java, where reflection exists.

**Comment by ericdallo** [comment]
It's completion for javascript, but the .query* is a method from the return of the js/document that return a Java object, am I wrong?

**Comment by bpringe** [comment]
I may not be getting what you are saying, but these are HTML DOM methods, essentially, from the `document` api: https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector

**Comment by ericdallo** [comment]
Sorry, I don't know cljs that much, my point is, I know that via cljs you can manipulate the DOM with some methods, but you are right, there is no java there, that clojurescript code it's transpiled to js later.
Anyway, not sure clj-kondo analysis would help here, WDYT @borkdude?

Otherwise, we would need to add some integration with a lib like clj-suitable, but would be big change to clojure-lsp probably

**Comment by borkdude** [comment]
@ericdallo Can you summarize what you would like from clj-kondo?

**Comment by ericdallo** [comment]
@borkdude, would be possible to have clj-kondo analysis for cljs DOM methods like from this [gif](https://github.com/rksm/clj-suitable/blob/master/doc/2019_07_22_suitable-figwheel.gif), I don't know how hard would be to implement that on clj-kondo if possible

**Comment by borkdude** [comment]
@ericdallo 

There is a stale PR here https://github.com/clj-kondo/clj-kondo/pull/879 to add arity info for JS interop calls. It's not merged and I'm not sure if will be but the work done in that issue could directly benefit clojure-lsp since the symbols resulting from that work could just be a built-in resource for JS completion in clojure-lsp. 

Similar approach can work for built-in Java stuff. Clj-kondo has some of this stuff, e.g. `(Thread/sleep)` will give an arity warning, but in practice I have never found this very useful. That said, the symbols can just be harvested (using reflection or javadoc) once and be built into clojure-lsp for completion.

**Comment by bpringe** [comment]
I was also thinking not just built-in, but also, for example in Calva we require things like `"vscode"`, and it would be nice if auto completions would show properties and methods on `vscode` when I type something like `(.| vscode)`. This would require dynamically analyzing the JS objects, and I don't know what that involves. Just dreaming :smile:. 

This isn't even a huge deal imo. It's just that when using JS I can type `vscode.` and see all props and methods, but using cljs I need to use the browser to look up docs, essentially, if I want to know what's available.

**Comment by darky** [comment]
Maybe it should be alternative clojurescript-lsp implementation, that communicate with tsserver directly and use all huge TypeScript types. Something like:
https://github.com/phil294/coffeesense/

**Comment by ericdallo** [comment]
Sounds more likely indeed @darky
